File: General/AttributeSetEditor/vb/AttributeSetEditor/MainForm.vb
Option Strict Off

Imports SolidEdgeCommunity
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Runtime.InteropServices
Imports System.Text
Imports System.Windows.Forms

Partial Public Class MainForm
    Inherits Form

    Private _application As SolidEdgeFramework.Application
    Private _document As SolidEdgeFramework.SolidEdgeDocument
    Private _documentEvents As SolidEdgeFramework.ISEDocumentEvents_Event

    Public Sub New()
        InitializeComponent()

        imageList.Images.Add(My.Resources.AttributeSets_16x16)
        imageList.Images.Add(My.Resources.AttributeSet_16x16)
        imageList.Images.Add(My.Resources.AttributeSetMissing_16x16)
    End Sub

    Private Sub MainForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
        Try
            UpdateTreeView()
        Catch
        End Try
    End Sub

    Private Sub MainForm_FormClosing(ByVal sender As Object, ByVal e As FormClosingEventArgs) Handles Me.FormClosing
        ReleaseDocument()
    End Sub

    Private Sub buttonRefresh_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonRefresh.Click
        UpdateTreeView()
    End Sub

    Private Sub UpdateTreeView()
        If _application Is Nothing Then
            Try
                _application = SolidEdgeUtils.Connect()
            Catch
                MessageBox.Show("Solid Edge is not running.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
                Return
            End Try
        End If

        If _document Is Nothing Then
            Try
                _document = DirectCast(_application.ActiveDocument, SolidEdgeFramework.SolidEdgeDocument)
            Catch
            End Try
        End If

        If _document IsNot Nothing Then
            Try
                _documentEvents = DirectCast(_document.DocumentEvents, SolidEdgeFramework.ISEDocumentEvents_Event)
                AddHandler _documentEvents.BeforeClose, AddressOf _documentEvents_BeforeClose
                AddHandler _documentEvents.SelectSetChanged, AddressOf _documentEvents_SelectSetChanged

                UpdateTreeView(_application.ActiveSelectSet)
            Catch ex As System.Exception
                _document = Nothing
                MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            End Try
        End If
    End Sub

    Private Sub UpdateTreeView(ByVal selectSet As SolidEdgeFramework.SelectSet)
        Dim rootNode = tvSelectSet.Nodes(0)
        rootNode.Nodes.Clear()
        lvAttributes.Items.Clear()

        For i As Integer = 1 To selectSet.Count
            Dim item As Object = Nothing
            Dim itemType As Type = Nothing

            Try
                item = selectSet.Item(i)
                itemType = SolidEdgeCommunity.Runtime.InteropServices.ComObject.GetType(item)
            Catch
            End Try

            If (item IsNot Nothing) AndAlso (itemType IsNot Nothing) Then
                Dim itemNode As New TreeNode(String.Format("Item {0} ({1})", i, itemType.FullName))
                Dim [property] = itemType.GetProperty("AttributeSets")

                If [property] IsNot Nothing Then
                    itemNode.ImageIndex = 1
                Else
                    itemNode.ImageIndex = 2
                End If

                itemNode.SelectedImageIndex = itemNode.ImageIndex
                itemNode.Tag = item

                rootNode.Nodes.Add(itemNode)
            End If
        Next i

        If rootNode.Nodes.Count > 0 Then
            tvSelectSet.SelectedNode = rootNode.Nodes(0)
        End If

        rootNode.ExpandAll()
    End Sub

    Private Sub UpdateListView()
        Dim attributeSets As SolidEdgeFramework.AttributeSets = Nothing
        Dim attributeSet As SolidEdgeFramework.AttributeSet = Nothing
        Dim attribute As SolidEdgeFramework.Attribute = Nothing

        lvAttributes.Items.Clear()
        buttonAddAttribute.Enabled = False
        buttonRemoveAttribute.Enabled = False

        If lvAttributes.Tag IsNot Nothing Then
            buttonAddAttribute.Enabled = True

            Dim selectSetItem As Object = lvAttributes.Tag
            attributeSets = selectSetItem.AttributeSets

            For i As Integer = 1 To attributeSets.Count
                attributeSet = attributeSets.Item(i)

                For j As Integer = 1 To attributeSet.Count
                    Dim lvItem = New ListViewItem()

                    attribute = attributeSet.Item(j)
                    lvItem.Text = attributeSet.SetName
                    lvItem.SubItems.Add(attribute.Name)
                    lvItem.SubItems.Add(attribute.Value.ToString())
                    lvItem.SubItems.Add(attribute.Value.GetType().FullName)

                    lvAttributes.Items.Add(lvItem)
                Next j
            Next i
        End If
    End Sub

    Private Sub _documentEvents_BeforeClose()
        ReleaseDocument()
    End Sub

    Private Sub _documentEvents_SelectSetChanged(ByVal SelectSet As Object)
        If Me.InvokeRequired Then
            Me.Invoke(New Action(Of SolidEdgeFramework.SelectSet)(AddressOf UpdateTreeView), DirectCast(SelectSet, SolidEdgeFramework.SelectSet))
            Return
        End If
    End Sub

    Private Sub tvSelectSet_AfterSelect(ByVal sender As Object, ByVal e As TreeViewEventArgs) Handles tvSelectSet.AfterSelect
        lvAttributes.Tag = e.Node.Tag
        UpdateListView()
    End Sub

    Private Sub buttonAddAttribute_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonAddAttribute.Click
        Dim attributeSets As SolidEdgeFramework.AttributeSets = Nothing
        Dim attributeSet As SolidEdgeFramework.AttributeSet = Nothing
        Dim attribute As SolidEdgeFramework.Attribute = Nothing

        Dim dialog = New AddAttributeDialog()

        If dialog.ShowDialog(Me) = System.Windows.Forms.DialogResult.OK Then
            Dim setName = dialog.SetName
            Dim attributeName = dialog.AttributeName
            Dim attributevalue = dialog.AttributeValue
            Dim selectSetItem As Object = lvAttributes.Tag

            attributeSets = selectSetItem.AttributeSets

            Try
                attributeSet = attributeSets.Item(setName)
            Catch
                attributeSet = attributeSets.Add(setName)
            End Try

            attribute = attributeSet.Add(attributeName, SolidEdgeFramework.AttributeTypeConstants.seStringUnicode)
            attribute.Value = attributevalue
        End If

        UpdateListView()
    End Sub

    Private Sub buttonRemoveAttribute_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonRemoveAttribute.Click
        Dim attributeSets As SolidEdgeFramework.AttributeSets = Nothing
        Dim attributeSet As SolidEdgeFramework.AttributeSet = Nothing

        Dim itemsToDelete As New List(Of ListViewItem)()

        For Each item As ListViewItem In lvAttributes.SelectedItems
            If item.Selected Then
                itemsToDelete.Add(item)
            End If
        Next item

        Dim selectSetItem As Object = lvAttributes.Tag

        For Each item In itemsToDelete
            Dim setName = item.Text
            Dim attributeName = item.SubItems(1).Text

            attributeSets = selectSetItem.AttributeSets
            attributeSet = attributeSets.Item(setName)
            attributeSet.Remove(attributeName)

            If attributeSet.Count = 0 Then
                attributeSets.Remove(attributeSet.SetName)
            End If

            lvAttributes.Items.Remove(item)
        Next item
    End Sub

    Private Sub lvAttributes_ItemSelectionChanged(ByVal sender As Object, ByVal e As ListViewItemSelectionChangedEventArgs) Handles lvAttributes.ItemSelectionChanged
        If e.Item IsNot Nothing Then
            buttonRemoveAttribute.Enabled = True
        Else
            buttonRemoveAttribute.Enabled = False
        End If
    End Sub

    Private Sub exitToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs) Handles exitToolStripMenuItem.Click
        Close()
    End Sub

    Private Sub ReleaseDocument()
        If _documentEvents IsNot Nothing Then
            RemoveHandler _documentEvents.SelectSetChanged, AddressOf _documentEvents_SelectSetChanged
            RemoveHandler _documentEvents.BeforeClose, AddressOf _documentEvents_BeforeClose

            Marshal.FinalReleaseComObject(_documentEvents)
            _documentEvents = Nothing
        End If

        If _document IsNot Nothing Then
            Marshal.FinalReleaseComObject(_document)
            _document = Nothing
        End If
    End Sub
End Class

File: General/AttributeSetEditor/cs/AttributeSetEditor/MainForm.cs
using SolidEdgeCommunity;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Windows.Forms;

namespace AttributeSetEditor
{
    public partial class MainForm : Form
    {
        private SolidEdgeFramework.Application _application;
        private SolidEdgeFramework.SolidEdgeDocument _document;
        private SolidEdgeFramework.ISEDocumentEvents_Event _documentEvents;

        public MainForm()
        {
            InitializeComponent();

            imageList.Images.Add(AttributeSetEditor.Properties.Resources.AttributeSets_16x16);
            imageList.Images.Add(AttributeSetEditor.Properties.Resources.AttributeSet_16x16);
            imageList.Images.Add(AttributeSetEditor.Properties.Resources.AttributeSetMissing_16x16);
        }
        
        private void MainForm_Load(object sender, EventArgs e)
        {
            try
            {
                UpdateTreeView();
            }
            catch
            {
            }
        }

        private void MainForm_FormClosing(object sender, FormClosingEventArgs e)
        {
            ReleaseDocument();
        }

        private void buttonRefresh_Click(object sender, EventArgs e)
        {
            UpdateTreeView();
        }

        private void UpdateTreeView()
        {
            if (_application == null)
            {
                try
                {
                    _application = SolidEdgeUtils.Connect();
                }
                catch
                {
                    MessageBox.Show("Solid Edge is not running.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return;
                }
            }

            if (_document == null)
            {
                try
                {
                    _document = (SolidEdgeFramework.SolidEdgeDocument)_application.ActiveDocument;
                }
                catch
                {
                }
            }

            if (_document != null)
            {
                try
                {
                    _documentEvents = (SolidEdgeFramework.ISEDocumentEvents_Event)_document.DocumentEvents;
                    _documentEvents.BeforeClose += _documentEvents_BeforeClose;
                    _documentEvents.SelectSetChanged += _documentEvents_SelectSetChanged;

                    UpdateTreeView(_application.ActiveSelectSet);
                }
                catch (System.Exception ex)
                {
                    _document = null;
                    MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                }
            }
        }

        private void UpdateTreeView(SolidEdgeFramework.SelectSet selectSet)
        {
            var rootNode = tvSelectSet.Nodes[0];
            rootNode.Nodes.Clear();
            lvAttributes.Items.Clear();

            for (int i = 1; i <= selectSet.Count; i++)
            {
                dynamic item = null;
                Type itemType = null;

                try
                {
                    item = selectSet.Item(i);
                    itemType = SolidEdgeCommunity.Runtime.InteropServices.ComObject.GetType(item);
                }
                catch
                {
                }

                if ((item != null) && (itemType != null))
                {
                    TreeNode itemNode = new TreeNode(String.Format("Item {0} ({1})", i, itemType.FullName));
                    var property = itemType.GetProperty("AttributeSets");

                    if (property != null)
                    {
                        itemNode.ImageIndex = 1;
                    }
                    else
                    {
                        itemNode.ImageIndex = 2;
                    }
                    
                    itemNode.SelectedImageIndex = itemNode.ImageIndex;
                    itemNode.Tag = item;

                    rootNode.Nodes.Add(itemNode);
                }
            }

            if (rootNode.Nodes.Count > 0)
            {
                tvSelectSet.SelectedNode = rootNode.Nodes[0];
            }

            rootNode.ExpandAll();
        }

        private void UpdateListView()
        {
            SolidEdgeFramework.AttributeSets attributeSets = null;
            SolidEdgeFramework.AttributeSet attributeSet = null;
            SolidEdgeFramework.Attribute attribute = null;

            lvAttributes.Items.Clear();
            buttonAddAttribute.Enabled = false;
            buttonRemoveAttribute.Enabled = false;

            if (lvAttributes.Tag != null)
            {
                buttonAddAttribute.Enabled = true;

                dynamic selectSetItem = lvAttributes.Tag;
                attributeSets = selectSetItem.AttributeSets;

                for (int i = 1; i <= attributeSets.Count; i++)
                {
                    attributeSet = attributeSets.Item(i);

                    for (int j = 1; j <= attributeSet.Count; j++)
                    {
                        var lvItem = new ListViewItem();

                        attribute = attributeSet.Item(j);
                        lvItem.Text = attributeSet.SetName;
                        lvItem.SubItems.Add(attribute.Name);
                        lvItem.SubItems.Add(attribute.Value.ToString());
                        lvItem.SubItems.Add(attribute.Value.GetType().FullName);

                        lvAttributes.Items.Add(lvItem);
                    }
                }
            }
        }

        void _documentEvents_BeforeClose()
        {
            ReleaseDocument();
        }

        void _documentEvents_SelectSetChanged(object SelectSet)
        {
            if (this.InvokeRequired)
            {
                this.Invoke(new Action<SolidEdgeFramework.SelectSet>(UpdateTreeView), (SolidEdgeFramework.SelectSet)SelectSet);
                return;
            }   
        }

        private void tvSelectSet_AfterSelect(object sender, TreeViewEventArgs e)
        {
            lvAttributes.Tag = e.Node.Tag;
            UpdateListView();
        }

        private void buttonAddAttribute_Click(object sender, EventArgs e)
        {
            SolidEdgeFramework.AttributeSets attributeSets = null;
            SolidEdgeFramework.AttributeSet attributeSet = null;
            SolidEdgeFramework.Attribute attribute = null;

            var dialog = new AddAttributeDialog();
            
            if (dialog.ShowDialog(this) == System.Windows.Forms.DialogResult.OK)
            {
                var setName = dialog.SetName;
                var attributeName = dialog.AttributeName;
                var attributevalue = dialog.AttributeValue;

                dynamic selectSetItem = lvAttributes.Tag;

                attributeSets = selectSetItem.AttributeSets;
                
                try
                {
                    attributeSet = attributeSets.Item(setName);
                }
                catch
                {
                    attributeSet = attributeSets.Add(setName);
                }

                attribute = attributeSet.Add(attributeName, SolidEdgeFramework.AttributeTypeConstants.seStringUnicode);
                attribute.Value = attributevalue;
            }

            UpdateListView();
        }

        private void buttonRemoveAttribute_Click(object sender, EventArgs e)
        {
            SolidEdgeFramework.AttributeSets attributeSets = null;
            SolidEdgeFramework.AttributeSet attributeSet = null;

            List<ListViewItem> itemsToDelete = new List<ListViewItem>();

            foreach (ListViewItem item in lvAttributes.SelectedItems)
            {
                if (item.Selected)
                {
                    itemsToDelete.Add(item);
                }
            }

            dynamic selectSetItem = lvAttributes.Tag;

            foreach (var item in itemsToDelete)
            {
                var setName = item.Text;
                var attributeName = item.SubItems[1].Text;

                attributeSets = selectSetItem.AttributeSets;
                attributeSet = attributeSets.Item(setName);
                attributeSet.Remove(attributeName);

                if (attributeSet.Count == 0)
                {
                    attributeSets.Remove(attributeSet.SetName);
                }

                lvAttributes.Items.Remove(item);
            }
        }

        private void lvAttributes_ItemSelectionChanged(object sender, ListViewItemSelectionChangedEventArgs e)
        {
            if (e.Item != null)
            {
                buttonRemoveAttribute.Enabled = true;
            }
            else
            {
                buttonRemoveAttribute.Enabled = false;
            }
        }

        private void exitToolStripMenuItem_Click(object sender, EventArgs e)
        {
            Close();
        }

        private void ReleaseDocument()
        {
            if (_documentEvents != null)
            {
                _documentEvents.SelectSetChanged -= _documentEvents_SelectSetChanged;
                _documentEvents.BeforeClose -= _documentEvents_BeforeClose;

                Marshal.FinalReleaseComObject(_documentEvents);
                _documentEvents = null;
            }

            if (_document != null)
            {
                Marshal.FinalReleaseComObject(_document);
                _document = null;
            }
        }
    }
}
